LibraBFT/Impl/Consensus/SafetyRules/SafetyRules.agda

```diff
@@ -55,63 +55,100 @@ verifyAndUpdatePreferredRoundM quorumCert safetyData = do
-  verifyQcM (proposedBlock ^∙ bQuorumCert) ∙?∙ λ _ → do
-    validatorVerifier ← gets rmGetValidatorVerifier
-    pure (Block.validateSignature proposedBlock validatorVerifier) ∙?∙ λ _ →
-      verifyAndUpdatePreferredRoundM (proposedBlock ^∙ bQuorumCert) safetyData0 ∙?∙
-      constructAndSignVoteM-continue2 voteProposal validatorSigner proposedBlock
-
-constructAndSignVoteM-continue2 voteProposal validatorSigner proposedBlock safetyData =
-  verifyAndUpdateLastVoteRoundM (proposedBlock ^∙ bBlockData ∙ bdRound) safetyData ∙?∙ λ safetyData1 → do
+module constructAndSignVoteM-continue0 (voteProposal : VoteProposal) (validatorSigner : ValidatorSigner) where
+  step₀ : LBFT (FakeErr ⊎ Vote)
+  step₁ : SafetyData → LBFT (FakeErr ⊎ Vote)
+
+  proposedBlock = voteProposal ^∙ vpBlock
+  step₀ = do
+    safetyData0 ← use (lPersistentSafetyStorage ∙ pssSafetyData)
+    verifyEpochM (proposedBlock ^∙ bEpoch) safetyData0 ∙?∙ λ _ → step₁ safetyData0
+  step₁ safetyData0 = do
+      case (safetyData0 ^∙ sdLastVote) of λ where
+        (just vote) →
+          if ⌊ vote ^∙ vVoteData ∙ vdProposed ∙ biRound ≟ℕ (proposedBlock ^∙ bRound) ⌋
+            then ok vote
+            else constructAndSignVoteM-continue1 voteProposal validatorSigner proposedBlock safetyData0
+        nothing → constructAndSignVoteM-continue1 voteProposal validatorSigner proposedBlock safetyData0
+
+constructAndSignVoteM-continue0 = constructAndSignVoteM-continue0.step₀
+
+module constructAndSignVoteM-continue1
+  (voteProposal  : VoteProposal) (validatorSigner : ValidatorSigner)
+  (proposedBlock : Block)        (safetyData0     : SafetyData) where
+
+  step₀ : LBFT (FakeErr ⊎ Vote)
+  step₁ : LBFT (FakeErr ⊎ Vote)
+  step₂ : ValidatorVerifier → LBFT (FakeErr ⊎ Vote)
+  step₃ : LBFT (FakeErr ⊎ Vote)
+
+  step₀ =
+    verifyQcM (proposedBlock ^∙ bQuorumCert) ∙?∙ λ _ → step₁
+  step₁ = do
+      validatorVerifier ← gets rmGetValidatorVerifier
+      step₂ validatorVerifier
+  step₂ validatorVerifier =
+      pure (Block.validateSignature proposedBlock validatorVerifier) ∙?∙ λ _ → step₃
+  step₃ =
+        verifyAndUpdatePreferredRoundM (proposedBlock ^∙ bQuorumCert) safetyData0 ∙?∙
+        constructAndSignVoteM-continue2 voteProposal validatorSigner proposedBlock
+
+constructAndSignVoteM-continue1 = constructAndSignVoteM-continue1.step₀
+
+module constructAndSignVoteM-continue2 (voteProposal : VoteProposal) (validatorSigner : ValidatorSigner)
+                                       (proposedBlock : Block) (safetyData : SafetyData) where
+  step₀ : LBFT (FakeErr ⊎ Vote)
+  step₁ : SafetyData → LBFT (FakeErr ⊎ Vote)
+  step₂ : SafetyData → VoteData → LBFT (FakeErr ⊎ Vote)
+  step₃ : SafetyData → VoteData → Author → LedgerInfo → LBFT (FakeErr ⊎ Vote)
+
+  step₀ = verifyAndUpdateLastVoteRoundM (proposedBlock ^∙ bBlockData ∙ bdRound) safetyData ∙?∙ step₁
+
+  step₁ safetyData1 = do
     lSafetyData ∙= safetyData1
```

Comment:
I've added a `TODO-1` about this.

Reply:
I don't think we have decided yet, either (so I suppose that's a quorum for indecision, ha).I am for a naming convention in the form of `pssSafetyData-rm, so the name of the lens followed by the instance.

Alternatively, with a bit more work we can use Agda's instance arguments, keeping `pssSafetyData` as-is and redefining `∙=` so that it searches for evidence of a lens from `RoundManager` to `PersistentSafetyStorage`.

Reply:
I agree with the naming convention in the short term.  I think @haroldcarr had some visceral reaction to the spelling.  Let's provoke him by doing _something_, and see how much he cares :smile:.  Can you do this in one of your upcoming PRs?